### ISSUE TITLE
Is MOTIP a class-agnostic model for tracking-by-detection task?

### ISSUE THREAD

**Issue by max-unfinity** [MCG-NJU/MOTIP#36]
Hi, thank you for amazing work!

Is MOTIP a class-agnostic model for tracking-by-detection task?
For example, if I have my Deformable DETR checkpoint trained on some marine dataset, can I use your pre-trained MOTIP models without any re-training or fine-tuning? Specifically, is the SeqDecoder module class-agnostic?

**Comment by HELLORPG** [comment]
I'm sorry for the delay in my responses. I'll be in the hospital for a while (maybe a month or more), so my replies might be slower than usual. Thanks for your understanding and patience.

Thanks for your interest in our work. 
**Currently, our MOTIP is not a class-agnostic model.** The reason is that our SeqDecoder and DETR are trained jointly. This means that the weights (checkpoints) of these two parts need to be used together. Specifically, our SeqDecoder needs the object features from DETR (query's output embedding), so consistency should be ensured. 
***However, this pipeline (ID prediction) can be further extended to a class-agnostic method easily.*** Here is a feasible method: decouple the detector (DETR) and the SeqDecoder parts. The detector only provides the bounding boxes (currently, we also use the output embeddings), and the SeqDecoder should independently extract the corresponding object features from the raw images based on the given bounding boxes (currently, we use the DETR output embeddings as mentioned above). **This way, these two parts can be connected solely through the detector's output boxes, without the need for consistent training and usage.** 
🤗 ***Currently, our design aims to minimize the engineering details and complex designs that need to be considered during the exploration process. Therefore, it may not be the most suitable for practical applications.*** However, I believe we DO demonstrate that our proposed method has significant potential. ***I am more than willing to help transition this work to a wide range of application scenarios or assist in future research.***

**Comment by MattLiutt** [comment]
Thanks for the excellent repo! So do you mean that we can separate detector and SeqDecoder during inference? For instance, I get a general usage detector (DETR or YOLO series), can I then do it like Sort those tracking-by-detection methods? Or we need to train both detector and SeqDecoder jointly and then do the inference? Clarify me if there's any misunderstanding! Thanks a lot!

**Comment by HELLORPG** [comment]
> So do you mean that we can separate detector and SeqDecoder during inference? For instance, I get a general usage detector (DETR or YOLO series), can I then do it like Sort those tracking-by-detection methods?

The codebase we provide in this repo does not support this feature. My above reply means that our thinking (ID prediction for target association) can be extended to the association-only model (similar to ReID methods) rather than the joint detection and association model (as we did). 

If you want to get an association-only tracking-by-detection model, you need to re-write your own code based on ours. Once you ensure that feature extraction in SeqDecoder is decoupled from the detector, then this SeqDecoder can be combined with any trained detector to become a tracking-by-detection method you mentioned (not tied to any specific detector). You can refer to work like [MASA](https://github.com/siyuanliii/masa)/[PuTR](https://github.com/chongweiliu/PuTR) for inspiration, where they trained a decoupled feature extractor. 

> Or we need to train both detector and SeqDecoder jointly and then do the inference?

For this repo, YES. Because the detector (Deformable DETR) also plays the role of feature extractor for SeqDecoder. Therefore, the detector and SeqDecoder are coupled together, trained together and inferred together. 

I hope this clarifies your concerns. Please let me know if you need additional details.

**Comment by MattLiutt** [comment]
Thanks for prompt response! Appreciated! 

> For this repo, YES. Because the detector (Deformable DETR) also plays the role of feature extractor for SeqDecoder. Therefore, the detector and SeqDecoder are coupled together, trained together and inferred together.

Just one last question, this repo used Deformable DETR as the detector as well feature extractor for SeqDecoder, is it feasible to replace it with other transformer?

Thanks so much!

**Comment by HELLORPG** [comment]
> is it feasible to replace it with other transformer?

Yes, of course. In this repo, we also provide MOTIP-DAB-Deformable-DETR except the default MOTIP-Deformable-DETR, as reported in [dancetrack results](https://github.com/MCG-NJU/MOTIP?tab=readme-ov-file#dancer-dancetrack). 

Specifically, you can refer to the following code to use your own transformer detector (`self.detr`) and the corresponding criterion function (`self.detr_criterion`):

https://github.com/MCG-NJU/MOTIP/blob/1dda4c49b4e5df1bac6de2827d7ca780b48d8853/models/motip.py#L91-L100

Additionally, we need to make some modifications to the return values of the DETR detector to ensure it returns the target features (`output embeddings`):

https://github.com/MCG-NJU/MOTIP/blob/1dda4c49b4e5df1bac6de2827d7ca780b48d8853/models/deformable_detr/deformable_detr.py#L191-L194

**Comment by HELLORPG** [comment]
**In summary, extending the MOTIP approach into a classic-independent appearance matcher should be an engineering problem.** If suitable requirements and resources become available in the future, we will consider implementing this in an extended version of this paper.

I have closed this issue until someone wishes to revisit or engage in a deeper discussion.
